### PR TITLE
Expose existing bitmap aggregate indexes to Cascades correctly

### PR DESCRIPTION
**Pull request by hatyo**
This enables reading an existing `KeyExpression` of a bitmap index and translates it into a match candidate that mimics an equivalent SQL aggregate query:

```sql
select bitmap_construct_agg(bitmap_bit_position(x)) as bitmap, bitmap_bucket_offset(x) as offset 
from bitmapIdx
group by bitmap_bucket_offset(x)
```

This is done by having a special implementation of `AggregateIndexExpansionVisitor` called `BitmapAggregateIndexExpansionVisitor` that does more or less the following:
- wrap the (singular) aggregated (field) `x` value to become `bitmap_construct_agg(bitmap_bit_position(x))`.
- add an _extra_ grouping column representing the implicit grouping column for the bitmap-bucket-offset, i.e. `bitmap_bucket_offset(x)` added by the index maintainer.
- expose the extra grouping column as a `Placeholder` to be used in `select-having` so we can filter on it, similar to other grouping columns.

**Note**: the actual size of the bucket is read from the index option `IndexOptions.BITMAP_VALUE_ENTRY_SIZE_OPTION`.

This exposed a limitation in `AggregateIndexMatchCandidate`'s construction of the index-entry to record, since it wasn't accounting the extra bitmap-bucket-offset group created by the index maintainer, and was creating an incorrect `Copier` that crashes when attempting to read from the index. I fixed it by informing the index-entry to record that we do have an _extra_ group when reading a bitmap index.

### CODE REVIEW

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/GroupByTest.java [hunk 1]
```diff
@@ -289,7 +293,7 @@ void testBitMapWithBitMapIndex() {
                 Optional.empty(),
                 IndexQueryabilityFilter.TRUE,
                 EvaluationContext.empty()).getPlan();
-        assertBitMapResult(hook, plan, 1250);
+        assertBitMapResult(hook, plan, 1);
```
Comment (pengpeng-lu) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/GroupByTest.java:299]: I think this should be `IndexOptions.BITMAP_VALUE_ENTRY_SIZE_OPTION / 8`?

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/GroupByTest.java [hunk 2]
```diff
@@ -413,7 +417,8 @@ protected RecordMetaDataHook setupHookAndAddData(final boolean addIndex, final b
                     metaDataBuilder.addIndex("MySimpleRecord", new Index("AggIndex", field("num_value_3_indexed").groupBy(field("num_value_2")), IndexTypes.SUM));
                 }
                 if (addBitMapIndex) {
-                    metaDataBuilder.addIndex("MySimpleRecord", new Index("BitMapIndex", modExpression(field("num_value_2"), bucketSize).groupBy(field("str_value_indexed"), bitBucketExpression(field("num_value_2"), bucketSize)), IndexTypes.BITMAP_VALUE));
+                    //metaDataBuilder.addIndex("MySimpleRecord", new Index("BitMapIndex", modExpression(field("num_value_2"), bucketSize).groupBy(field("str_value_indexed"), bitBucketExpression(field("num_value_2"), bucketSize)), IndexTypes.BITMAP_VALUE));
+                    metaDataBuilder.addIndex("MySimpleRecord", new Index("BitMapIndex", field("num_value_2").groupBy(field("str_value_indexed")), IndexTypes.BITMAP_VALUE, ImmutableMap.of(IndexOptions.BITMAP_VALUE_ENTRY_SIZE_OPTION, "4")));
```
Comment (pengpeng-lu) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/GroupByTest.java]: "4" -> bitBucketSize?

File: .idea/qaplug_profiles.xml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (normen662) [inline .idea/qaplug_profiles.xml]: please remove this file.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/AggregateIndexExpansionVisitor.java [hunk 4]
```diff
@@ -71,16 +71,16 @@
 public class AggregateIndexExpansionVisitor extends KeyExpressionExpansionVisitor
                                             implements ExpansionVisitor<KeyExpressionExpansionVisitor.VisitorState> {
     @Nonnull
-    private static final Supplier<Map<String, BuiltInFunction<? extends Value>>> aggregateMap = Suppliers.memoize(AggregateIndexExpansionVisitor::computeAggregateMap);
+    static final Supplier<Map<String, BuiltInFunction<? extends Value>>> aggregateMap = Suppliers.memoize(AggregateIndexExpansionVisitor::computeAggregateMap);
 
     @Nonnull
-    private final Index index;
+    protected final Index index;
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/AggregateIndexExpansionVisitor.java:77]: I started using the pattern of making them `private` but giving them `protected` accessors. Doesn't really matter here.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/AggregateIndexExpansionVisitor.java [hunk 5]
```diff
@@ -130,8 +130,11 @@ public MatchCandidate expand(@Nonnull final Supplier<Quantifier.ForEach> baseQua
         // 2. create a GROUP-BY expression on top.
         final var groupByQun = constructGroupBy(selectWhereQunAndPlaceholders.getLeft(), baseExpansion);
 
+        final var placeholders = ImmutableList.<Placeholder>builder().addAll(selectWhereQunAndPlaceholders.getRight())
+                .addAll(groupByQun.getRight()).build();
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/AggregateIndexExpansionVisitor.java]: `groupByQun` is not really a qun. Can you rename?

**Review by normen662 — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/AggregateIndexExpansionVisitor.java (hunk 5 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/AggregateIndexExpansionVisitor.java]: Done.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/GroupByTest.java (hunk 1 above)
Comment (hatyo) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/GroupByTest.java:299]: The returned bitmap size (in bits) is determined in the constructor of `BitmapValueIndexMaintainer`:

https://github.com/pengpeng-lu/fdb-record-layer/blob/0d54f1e423ec1807e72460c229846654ab786e32/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/BitmapValueIndexMaintainer.java#L100-L101

So, according to the test, it should be `4`. Now, the actual byte array size is determined as the following:

https://github.com/pengpeng-lu/fdb-record-layer/blob/0d54f1e423ec1807e72460c229846654ab786e32/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/BitmapValueIndexMaintainer.java#L203

So, we should get `1` byte back in return, which is exactly correct.

**Comment by hatyo** [comment]
I think the PR is ready to merge in [pengpeng-lu:bitmap_func](https://github.com/pengpeng-lu/fdb-record-layer/tree/bitmap_func) branch. Few notes:

- I added some tests for verifying that we can still use a bitmap aggregate index that has zero explicit groups, i.e. the only group is contains, is the _implicitly_ defined group that is semantically equivalent to `bitmap_bucket_offset(x)`.
- The test `testBitMapWithStreamAggregation` is not passing, I think it is also not passing in the outstanding PR #2744. It looks like it is only failing because the size of the bitmap bucket is larger so the results are different from what's written in the expected results. The method `assertBitMapResult` should be adapted accordingly, or made more intelligent such that it recreates the expected result according to the expected bit bucket size.

**Review by pengpeng-lu — APPROVED** [review]

**Pull request merged**